fix: Allow narrowing a `Relationship` by its `target-type` and a `Collection` by its `entity-type`

mwiencek opened the pull request:
TypeScript has the concept of discriminating unions by a shared field name. For example:

```TypeScript
declare var relations: Relationship[];
for (const r of relations) {
  switch (r["target-type"]) {
    case "area":
      r.area.name;
      break;
    case "release_group":
      r.release_group.title;
      break;
  }
}
```

On the `main` branch, this currently errors:

    error: TS18048 [ERROR]: 'r.area' is possibly 'undefined'.
          r.area.name;
          ~~~~~~
        at file:///home/michael/code/musicbrainz-ts/test/data/lookup.ts:1180:7
   
    TS18048 [ERROR]: 'r.release_group' is possibly 'undefined'.
          r.release_group.title;
          ~~~~~~~~~~~~~~~
        at file:///home/michael/code/musicbrainz-ts/test/data/lookup.ts:1183:7
   
    Found 2 errors.

With the changes in this commit, the same code type-checks. (It also blocks you from accessing `r.release_group` under the `"area"` branch, or `r.area` under the `"release_group"` branch, etc.).

The only way I could get this to work was by partially re-enabling the distributivity that was disabled in 72e3ba55688793fc4e26d02d402b1763243a77fa; I kept it disabled for `string` and `number`, which I believe are the only two primitives *contained in arrays* we need to be concerned about.

@kellnerd If you agree to this patch (and it doesn't cause any issues I didn't anticipate), a similar patch could be applied to collections.

mwiencek committed:
fix: Allow narrowing a `Relationship` by its `target-type`

TypeScript has the concept of discriminating unions by a shared field name. For
example:

```TypeScript
declare var relations: Relationship[];
for (const r of relations) {
  switch (r["target-type"]) {
    case "area":
      r.area.name;
      break;
    case "release_group":
      r.release_group.title;
      break;
  }
}
```

On the `main` branch, this currently errors:

    error: TS18048 [ERROR]: 'r.area' is possibly 'undefined'.
          r.area.name;
          ~~~~~~
        at file:///home/michael/code/musicbrainz-ts/test/data/lookup.ts:1180:7

    TS18048 [ERROR]: 'r.release_group' is possibly 'undefined'.
          r.release_group.title;
          ~~~~~~~~~~~~~~~
        at file:///home/michael/code/musicbrainz-ts/test/data/lookup.ts:1183:7

    Found 2 errors.

With the changes in this commit, the same code type-checks. (It also blocks you
from accessing `r.release_group` under the `"area"` branch, or `r.area` under
the `"release_group"` branch, etc.).

The only way I could get this to work was by partially re-enabling the
distributivity that was disabled in 72e3ba55688793fc4e26d02d402b1763243a77fa; I
kept it disabled for `string` and `number`, which I believe are the only two
primitives *contained in arrays* we need to be concerned about.

A similar patch could be applied to collections.

mwiencek commented:
P.S. The diff is best viewed with "Hide whitespace" enabled.

P.P.S. I wonder if it would even be possible to map the included `$entity-rels` inc parameters directly into the `Relationship<...>` type parameter; e.g., `"artist-rels" | "event-rels"` would map to `SubQuery<Relationship<"artist" | "event">[], ...>`.

kellnerd commented on api_types.ts:
```diff
@@ -487,51 +487,48 @@ export interface GenreTag extends GenreUserTag {
+      /** Credited name of the target entity, can be empty. */
+      "target-credit": string;
+    }
+    & DatePeriod
```
```suggestion
    & DatePeriod;
```
Just making the formatter (`deno fmt`) happy :)

mwiencek replied:
Thank you, I didn't know about `deno fmt` until now!

kellnerd commented on api_types.ts:
```diff
@@ -487,51 +487,48 @@ export interface GenreTag extends GenreUserTag {
+};
+
+export type Relationship<
+  TargetType extends RelatableEntityType = RelatableEntityType
```
```suggestion
  TargetType extends RelatableEntityType = RelatableEntityType,
```

kellnerd commented on api_includes.ts:
```diff
@@ -80,7 +80,9 @@ type UnwrapData<Data, Include extends IncludeParameter> =
-      ? [Item] extends [object] ? WithIncludes<Item, Include>[] : Data
+      ? [Item] extends [string] ? Item[]
+      : [Item] extends [number] ? Item[]
+      : Item extends object ? WithIncludes<Item, Include>[] : Data
```
```suggestion
      ? [Item] extends [string | number | undefined] ? Item[]
      : Item extends object ? WithIncludes<Item, Include>[]
      : Item[]
```
We can combine the two cases here and also add `undefined` in case a union type includes it.
Cool that it is finally possible to state the type as `Item[]` which is clearer than just `Data`, I think this was not possible before turning off the distributivity.
While I don't always agree with the formatters behavior for these complex types (and sneaked in some comments at the right spot to force line breaks where I want to have them), it makes sense to me to have the fallback case (which should not happen for our use case) on its own line.

kellnerd reviewed (commented):
> If you agree to this patch (and it doesn't cause any issues I didn't anticipate), a similar patch could be applied to collections.

I definitely agree and wonder why I didn't came up with this approach but with the more complex and less accurate `IfUnionType` helper 😊
Collections should definitely also follow this approach, do you want to do that as part of this PR?

> I wonder if it would even be possible to map the included $entity-rels inc parameters directly into the `Relationship<...>` type parameter; e.g., `"artist-rels" | "event-rels"` would map to `SubQuery<Relationship<"artist" | "event">[], ...>`.

At least I hope that this is possible, otherwise I probably would not have introduced the generic parameter. But so far I have not come up with a clever solution to achieve this.
Defining a helper to map relationship includes to entity types ~should be~ is easy:
```ts
type ExtractRelTargetType<Include extends IncludeParameter> = Include extends
  `${infer Type extends RelatableEntityType}-rels` ? Type : never;
```

The problem is that at the point where `SubQuery` is used, we know nothing about the given include parameters which will later be specified for `WithIncludes`.
I guess we would have to add a generic parameter for the given includes to all entity types and this makes everything quite ugly... but I don't know if there is a solution which is able to avoid this.

Since I had already thought about only exposing type aliases instead of the entity types and helpers in the final version, I think this might be the way to go:
```ts
// Simple version
export type ArtistWith<Include extends IncludeParameter> = WithIncludes<
  Artist,
  Include
>;

// Version with restricted include values
export type ArtistWith<Include extends CollectIncludes<Artist>> = WithIncludes<
  Artist,
  Include
>;

// Potential version with restricted include values and relationship types
export type ArtistWith<Include extends CollectIncludes<Artist>> = WithIncludes<
  Artist<Include>,
  Include
>; 
```

P.S. If we are going that route, we could also rename the internal types to something else, maybe `$Artist` and use the current names for the public types and have the includes default to none (`Artist<Include = never>`). What do you think?

mwiencek committed:
fix: Allow narrowing a `Collection` by its `entity-type`

This follows the same implementation as
fb4b73f850cb78e5ac8f096e3bfa8cc029185479.

Removes the now-unused `IfUnionType` utility.

mwiencek commented:
> Collections should definitely also follow this approach, do you want to do that as part of this PR?

Sure, I pushed a second commit for that.

> At least I hope that this is possible, otherwise I probably would not have introduced the generic parameter. But so far I have not come up with a clever solution to achieve this.
> Defining a helper to map relationship includes to entity types should be is easy:

Your helper type matches what I had imagined too. :smiley: 

> The problem is that at the point where SubQuery is used, we know nothing about the given include parameters which will later be specified for WithIncludes.
> I guess we would have to add a generic parameter for the given includes to all entity types and this makes everything quite ugly... but I don't know if there is a solution which is able to avoid this.

Unfortunately, I think you are right that adding another generic is the only way. And it might be ugly...

When you had mentioned issues with recursion before, I even figured you might have to eventually add *another* generic onto entities that provides information about the root request type. But you've been able to avoid that so far.

> Since I had already thought about only exposing type aliases instead of the entity types and helpers in the final version, I think this might be the way to go:
> [...]
> P.S. If we are going that route, we could also rename the internal types to something else, maybe $Artist and use the current names for the public types and have the includes default to none (Artist<Include = never>). What do you think?

That sounds sensible to me. The exposed types will be simpler and have a minimal default that doesn't enforce any includes. Also, I imagine that in cases where the inc parameters are *string literals*, the library could be smart enough to instantiate the response entity type directly. :exploding_head: So the API user generally won't have to instantiate these types directly.

kellnerd reviewed (approved):
Thank you, looking good to me!

> Also, I imagine that in cases where the inc parameters are string literals, the library could be smart enough to instantiate the response entity type directly. 🤯 So the API user generally won't have to instantiate these types directly.

That is already happening using the 14 overloads of the `lookup` method:

```ts
lookup<Include extends CollectIncludes<Artist> = never>(
  entityType: "artist",
  mbid: MBID,
  inc?: Include[],
): Promise<WithIncludes<Artist, Include>>;
// ...
lookup(
  entityType: EntityType,
  mbid: MBID,
  inc?: string[],
): Promise<EntityBase>;
```

The only thing which I haven't figured out yet, if/how I can prevent TS from falling back to the last overload as soon as an unsupported include parameter is specified for the given entity type...
That means the `inc` parameter has no auto-complete and you only see from the method's return type (displayed in the auto-complete tooltip) whether you have typed a valid include parameter.
I mean I could get autocomplete for `inc` by removing the last overload, but this makes the method impossible to use for an unknown entity type (as the testdata download script does).
